server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/DefaultMetaRepository.java

```diff
@@ -68,307 +55,147 @@ public class DefaultMetaRepository extends RepositoryWrapper implements MetaRepo
 
-            for (JsonNode m : mirrorsJson) {
-                final MirrorConfig c = Jackson.treeToValue(m, MirrorConfig.class);
-                if (c == null) {
-                    throw new RepositoryMetadataException(PATH_MIRRORS + " contains null.");
-                }
-                mirrors.addAll(c.toMirrors(parent(), credentials));
+        int index = 0;
+        for (JsonNode m : mirrorsJson) {
+            final MirrorConfig c = Jackson.treeToValue(m, MirrorConfig.class);
+            if (c == null) {
+                throw new RepositoryMetadataException(PATH_MIRRORS + " contains null.");
+            }
+            final Mirror mirror = c.toMirror(parent(), credentials, index++);
+            if (mirror != null) {
+                mirrors.add(mirror);
             }
-
-            return mirrors.build();
-        } catch (RepositoryMetadataException e) {
-            throw e;
-        } catch (Exception e) {
-            throw new RepositoryMetadataException("failed to load the mirror configuration", e);
         }
+        return mirrors.build();
     }
 
-    private static List<MirrorCredential> loadCredentials(Map<String, Entry<?>> entries) throws Exception {
+    private static List<MirrorCredential> parseCredentials(Map<String, Entry<?>> entries)
+            throws JsonProcessingException {
         final Entry<?> e = entries.get(PATH_CREDENTIALS);
         if (e == null) {
-            return Collections.emptyList();
+            return ImmutableList.of();
         }
 
         final JsonNode credentialsJson = (JsonNode) e.content();
         if (!credentialsJson.isArray()) {
             throw new RepositoryMetadataException(
                     PATH_CREDENTIALS + " must be an array: " + credentialsJson.getNodeType());
         }
-
-        if (credentialsJson.size() == 0) {
-            return Collections.emptyList();
+        if (credentialsJson.isEmpty()) {
+            return ImmutableList.of();
         }
 
         final ImmutableList.Builder<MirrorCredential> builder = ImmutableList.builder();
+        int index = 0;
         for (JsonNode c : credentialsJson) {
             final MirrorCredential credential = Jackson.treeToValue(c, MirrorCredential.class);
             if (credential == null) {
                 throw new RepositoryMetadataException(PATH_CREDENTIALS + " contains null.");
             }
+            assert credential instanceof AbstractMirrorCredential;
+            ((AbstractMirrorCredential) credential).setIndex(index);
             builder.add(credential);
+            index++;
         }
 
         return builder.build();
     }
-
-    @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
-    @JsonSubTypes({
-            @Type(value = SingleMirrorConfig.class, name = "single"),
-            @Type(value = MultipleMirrorConfig.class, name = "multiple")
-    })
-    private abstract static class MirrorConfig {
```

Comment:
`MultipleMirrorConfig` is removed and `SingleMirrorConfig` is migrated to `MirrorConfig.java`

Reply:
I'd feel better if we migrate the internally used `MultipleMirrorConfig` before merging this PR (or at least before the next release so that the next release driver isn't suprised)